### ISSUE TITLE
It fails the compilation

### ISSUE THREAD

**Issue by christenbc** [zhanghanduo/cubicle_detect#2]
<img width="505" alt="Sketch" src="https://user-images.githubusercontent.com/45597426/55055271-2d1b5700-506b-11e9-8064-74e7e84b8059.png">

After installing all the 3 required dependencies and compiling the catkin_make
`catkin_make -DCMAKE_BUILD_TYPE=Release`
I don't know why, I do get this error. Is it possible to run this package without CUDA or any GPU?

**Comment by zhanghanduo** [comment]
Yeah the problem is that you don't have CUDA so the project_name_lib cannot be build thus cannot be found. I can make updates enabling the code without CUDA or any GPU, but the problem is that your program will be really slow. The yolo module and sgm module all require high calculation capability hardware otherwise it will be a pain running this code. If you really wanna run codes like this, I suggest you find a computer with good gpu.

**Comment by zhanghanduo** [comment]
It seems you are using virtual machine to run ubuntu. Recently I am busy with other things. I guess I will update the code using CPU mode next week. Currently I suggest you to use another good repo: https://github.com/leggedrobotics/darknet_ros. This repo does not have stereo information, only running YOLO darknet under ROS. It can be run with solely CPU. Good luck!

**Comment by christenbc** [comment]
All right. So with a computer with a GPU is enough or in addition requires of CUDA?
How many fps do you think is going to run without GPU?
The location output data of your package are bounding boxes in 3D? what does it mean? that there are cubic boxes or there are 2D boxes located located in space? if so, the depth location of such bounding boxes is taken by the depth value of the central pixel or it does an average of the depth pixels in the bounding box?

**Comment by zhanghanduo** [comment]
1. We need CUDA. WIthout GPU might be really slow (I guess might be several seconds). Please refer to this pose: https://github.com/pjreddie/darknet/issues/872. I recommend you to use OpenCV directly.
2. Output format
Just refer to the [obstacle_msgs](https://github.com/zhanghanduo/obstacle_msgs): 
point3 centerPos    # 3D bounding box center position (meter)
float32 diameter    # diameter (meter)
float32 height
string classes
float32 probability
uint32 xmin         # 2D bounding box edge coordinate
uint32 ymin
uint32 xmax
uint32 ymax
It has both 2D with boundaries and 3D information with center location. We use median filter to make sure the center of the box is consistent, not noise or outlier.
However, if you want to average over the bounding box it is also OK. You just need to read the source code and do some changes.

**Comment by christenbc** [comment]
All right, thanks for the info.
Still I have some doubts regarding this:

> point3 centerPos # 3D bounding box center position (meter)
> float32 diameter # diameter (meter)
> float32 height

The first one means the position X, Y, X coordinates of the bounding box anchors, right?
What are the other 2 for?

And an important question, how can I use this package by using as input the RGB image and point cloud instead of "/camera_reading ([sensor_msgs/Image])" which I assume is /camera/depth/image_rect_raw?
Because /camera/depth/image_rect_raw is not working well with my camera.

**Comment by zhanghanduo** [comment]
Diameter and height means the 3D dimension or size of the bounding box. 
For your second question, I don't understand what do you mean by RGB image and point cloud input. This is a stereo camera package, you can refer to https://github.com/zhanghanduo/cubicle_detect/blob/master/darknet_ros/launch/kitti.launch to see the required sensor topic is left, right image_rect and camera_info. So you don't need point cloud or depth image. It uses built-in sgm method to calculate the depth automatically. If you already have point cloud, you have to change the source code accordingly.

**Comment by christenbc** [comment]
The point is that my camera just produces this set of topics:
```
/camera_floor/background
/camera_floor/camera_info
/camera_floor/driver/color/camera_info
/camera_floor/driver/color/image_raw
/camera_floor/driver/color/image_raw/compressed
/camera_floor/driver/color/image_raw/compressed/parameter_descriptions
/camera_floor/driver/color/image_raw/compressed/parameter_updates
/camera_floor/driver/depth/camera_info
/camera_floor/driver/depth/color/points
/camera_floor/driver/depth/image_rect_raw
/camera_floor/driver/depth/image_rect_raw/compressed
/camera_floor/driver/depth/image_rect_raw/compressed/parameter_descriptions
/camera_floor/driver/depth/image_rect_raw/compressed/parameter_updates
/camera_floor/driver/extrinsics/depth_to_color
/camera_floor/driver/parameter_descriptions
/camera_floor/driver/parameter_updates
/camera_floor/filter/parameter_descriptions
/camera_floor/filter/parameter_updates
/camera_floor/filter/visualization_marker
/camera_floor/floor
/camera_floor/obstacles
/camera_floor/transform/parameter_descriptions
/camera_floor/transform/parameter_updates
/camera_top/background
/camera_top/obstacles
```
This means that 
> /camera_floor/driver/depth/image_rect_raw 

is the only depth image topic, that is, there are no left nor right separated images.
It produces also: 

> /camera_floor/driver/depth/color/points
> /camera_floor/driver/color/image_raw

And regarding the shapes of the bounding boxes, then they are cubes if I understand well, in such case, how does your algorithm select the depth of such bounding box? Considering all depths the same in every class or depending on the class? The center of such cube aims the center of gravity of the detected object?

**Comment by zhanghanduo** [comment]
Not that complicated. The center of the 3D cube is just the center of the 2D rectangle. Note that in case of the center is outlier or noise, we use a 3x3 median filter to find out the center depth. We just assume for one object the depths of all points are the same. 

For your case, you have to modify my code to work on your camera. In darknet_ros/src/YoloObjectDetector.cu, the function getDepth() is used to generate disparity map. Since you already have the depth map, you can easily replace it with your depth map using callback function.

**Comment by chbloca** [comment]
All right. Now I managed to install the NVidia drivers and CUDA.
I install openCV with this guide https://gist.github.com/raulqf/a3caa97db3f8760af33266a1475d0e5e
and when trying to compile your package, after installing the other dependencies, this is what happens:
![WhatsApp Image 2019-03-31 at 1 21 47 AM](https://user-images.githubusercontent.com/45597426/55282816-d5c00400-5354-11e9-906d-a92a32288b6d.jpeg)
Do you know what might be the cause?

**Comment by christenbc** [comment]
Ok, not everything compiled properly. After downloading the weights and configuring properly the config file, then I `roslaunch cubicle_detect yolo_v3.launch` but I have no idea why the program I calling yolo_bdd... weights since I defined that I wanted to use yolov3.weights and yolov3.cfg

![1](https://user-images.githubusercontent.com/45597426/55401407-5c453300-5559-11e9-85bb-cfbc47bf4e04.jpg)
![2](https://user-images.githubusercontent.com/45597426/55401411-5d766000-5559-11e9-9858-5f1fb7a4e610.jpg)

**Comment by zhanghanduo** [comment]
Maybe the program bug? You can dive into the darknet_ros/src/YoloObjectDetector.cu, edit the lines:   nodeHandle_.param("yolo_model/weight_file/name", and nodeHandle_.param("yolo_model/config_file/name", configModel, std::string("yolo_bdd_c1.cfg"));.

**Comment by christenbc** [comment]
it is a very weird error since im doing a launch of yolo_v3, not the yolo_v3_bdd

**Comment by christenbc** [comment]
Yes, problem solved.
Now the last issue is that keeps waiting for the image when I am feeding it with my /camera_floor/driver/depth/image_rect_raw from a [Intel RealSense camera](https://click.intel.com/intelr-realsensetm-depth-camera-d435.html). Maybe this camera is not compatible with your package

**Comment by zhanghanduo** [comment]
I think if you are familiar with ROS you can change your callback function to adapt to your case.

**Comment by christenbc** [comment]
It should work with the topic I provided since it is the type of 16UC1.
I'm not sure if I should use another topic..
Here there is a full [list](https://github.com/IntelRealSense/librealsense/issues/3218):
rostopic list yield the followings:
/camera/Motion_Module/parameter_descriptions
/camera/Motion_Module/parameter_updates
/camera/RGB_Camera/parameter_descriptions
/camera/RGB_Camera/parameter_updates
/camera/Stereo_Module/parameter_descriptions
/camera/Stereo_Module/parameter_updates
/camera/accel/imu_info
/camera/accel/sample
/camera/color/camera_info
/camera/color/image_raw
/camera/color/image_raw/compressed
/camera/color/image_raw/compressed/parameter_descriptions
/camera/color/image_raw/compressed/parameter_updates
/camera/depth/camera_info
/camera/depth/image_rect_raw
/camera/depth/image_rect_raw/compressed
/camera/depth/image_rect_raw/compressed/parameter_descriptions
/camera/depth/image_rect_raw/compressed/parameter_updates
/camera/extrinsics/depth_to_color
/camera/extrinsics/depth_to_infra1
/camera/extrinsics/depth_to_infra2
/camera/gyro/imu_info
/camera/gyro/sample
/camera/infra1/camera_info
/camera/infra1/image_rect_raw
/camera/infra1/image_rect_raw/compressed
/camera/infra1/image_rect_raw/compressed/parameter_descriptions
/camera/infra1/image_rect_raw/compressed/parameter_updates
/camera/infra2/camera_info
/camera/infra2/image_rect_raw
/camera/infra2/image_rect_raw/compressed
/camera/infra2/image_rect_raw/compressed/parameter_descriptions
/camera/infra2/image_rect_raw/compressed/parameter_updates
/camera/realsense2_camera_manager/bond
/diagnostics
/rosout
/rosout_agg
/tf_static

Besides, I am watching that your code is processing left and right images, but the input of your system is from just a single topic.
